src/net/server/mod.rs

```diff
@@ -35,31 +38,32 @@ pub enum MpcHelperServerError {
         "Request body length {body_len} is not aligned with size of the element {element_size}"
     )]
     WrongBodyLen { body_len: u32, element_size: usize },
-    #[error(transparent)]
-    BadPathString(#[from] PathRejection),
+    #[error("bad path: {0}")]
+    BadPathString(#[source] BoxError),
     #[error(transparent)]
     BodyAlreadyExtracted(#[from] axum::extract::rejection::BodyAlreadyExtracted),
     #[error(transparent)]
+    MissingExtension(#[from] axum::extract::rejection::ExtensionRejection),
+    #[error("out-of-order delivery of data: expected index {last_seen}, but found {next_seen}")]
+    OutOfOrder { last_seen: u32, next_seen: u32 },
+    #[error(transparent)]
     HyperError(#[from] hyper::Error),
     #[error("parse error: {0}")]
     SerdeError(#[from] serde_json::Error),
     #[error("could not forward messages: {0}")]
     SendError(BoxError),
 }
 
-/// [`From`] implementation for [`MpcServerError::SendError`].
-/// first call `to_string` so as to drop `T` from the [`MpcServerError`]
-impl<T> From<mpsc::error::SendError<T>> for MpcHelperServerError {
-    fn from(err: mpsc::error::SendError<T>) -> Self {
-        Self::SendError(err.to_string().into())
+impl MpcHelperServerError {
+    pub fn query_id_not_found(query_id: QueryId) -> MpcHelperServerError {
+        Self::BadPathString(format!("encountered unknown query id: {}", query_id.as_ref()).into())
     }
-}
 
-/// [`From`] implementation for [`MpcServerError::SendError`].
-/// first call `to_string` to as to drop `T` from the [`MpcServerError`]
-impl<T> From<tokio_util::sync::PollSendError<T>> for MpcHelperServerError {
-    fn from(err: tokio_util::sync::PollSendError<T>) -> Self {
-        Self::SendError(err.to_string().into())
+    pub fn out_of_order(last_seen: u32, next_seen: u32) -> Self {
```

Comment:
if that is not too much work, would be great to add channel id here.

Reply:
done